File: backend/src/routes/drafts.js
import express from 'express';
import { requireAuth } from '../middleware/auth.js';
import prisma from '../database.js';

const router = express.Router();

// Get all drafts for user
router.get('/', requireAuth, async (req, res) => {
  try {
    const drafts = await prisma.resumeDraft.findMany({
      where: { userId: req.user.id },
      orderBy: { updatedAt: 'desc' }
    });
    res.json({ status: 'success', data: drafts });
  } catch (error) {
    console.error('Error fetching drafts:', error);
    res.status(500).json({ error: { message: 'Failed to fetch drafts.' } });
  }
});

// Create new draft
router.post('/', requireAuth, async (req, res) => {
  try {
    const { title, templateId, sections } = req.body;
    const draft = await prisma.resumeDraft.create({
      data: {
        userId: req.user.id,
        title: title || 'Untitled Draft',
        templateId: templateId || 'classic',
        templateSettings: req.body.templateSettings || {},
        sections: sections || [],
        roleId: req.body.roleId || null,
        branchOfVersionId: req.body.branchOfVersionId || null
      }
    });
    res.status(201).json({ status: 'success', data: draft });
  } catch (error) {
    console.error('Error creating draft:', error);
    res.status(500).json({ error: { message: 'Failed to create draft.' } });
  }
});

// Get specific draft
router.get('/:id', requireAuth, async (req, res) => {
  try {
    const draft = await prisma.resumeDraft.findFirst({
      where: { id: req.params.id, userId: req.user.id }
    });
    if (!draft) return res.status(404).json({ error: { message: 'Draft not found.' } });
    res.json({ status: 'success', data: draft });
  } catch (error) {
    console.error('Error fetching draft:', error);
    res.status(500).json({ error: { message: 'Failed to fetch draft.' } });
  }
});

// Update/Autosave draft
router.patch('/:id', requireAuth, async (req, res) => {
  try {
    // Verify ownership
    const existing = await prisma.resumeDraft.findFirst({
      where: { id: req.params.id, userId: req.user.id }
    });
    if (!existing) return res.status(404).json({ error: { message: 'Draft not found.' } });

    const { title, templateId, templateSettings, sections, roleId } = req.body;
    
    // Create data object dynamically based on provided fields
    const dataToUpdate = { revision: { increment: 1 } };
    if (title !== undefined) dataToUpdate.title = title;
    if (templateId !== undefined) dataToUpdate.templateId = templateId;
    if (templateSettings !== undefined) dataToUpdate.templateSettings = templateSettings;
    if (sections !== undefined) dataToUpdate.sections = sections;
    if (roleId !== undefined) dataToUpdate.roleId = roleId;

    const draft = await prisma.resumeDraft.update({
      where: { id: req.params.id },
      data: dataToUpdate
    });
    res.json({ status: 'success', data: draft });
  } catch (error) {
    console.error('Error updating draft:', error);
    res.status(500).json({ error: { message: 'Failed to update draft.' } });
  }
});

export default router;

File: backend/app/routes/drafts.py
"""
ResumeIQ — Resume Builder Drafts & Autosave Routes (FastAPI)
"""

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.orm import Session
from typing import Optional, List, Dict, Any

from ..database import get_db, User, ResumeDraft
from ..middleware.auth import get_current_user

router = APIRouter(prefix="", tags=["Drafts"])

class CreateDraftRequest(BaseModel):
    title: Optional[str] = "Untitled Draft"
    templateId: Optional[str] = "classic"
    templateSettings: Optional[Dict[str, Any]] = None
    sections: Optional[List[Dict[str, Any]]] = None
    roleId: Optional[str] = None
    branchOfVersionId: Optional[str] = None

class UpdateDraftRequest(BaseModel):
    title: Optional[str] = None
    templateId: Optional[str] = None
    templateSettings: Optional[Dict[str, Any]] = None
    sections: Optional[List[Dict[str, Any]]] = None
    roleId: Optional[str] = None

@router.get("/drafts")
@router.get("/v1/drafts")
def list_drafts(
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    drafts = (
        db.query(ResumeDraft)
        .filter(ResumeDraft.userId == user.id)
        .order_by(ResumeDraft.updatedAt.desc())
        .all()
    )
    return {
        "status": "success",
        "data": [
            {
                "id": d.id,
                "title": d.title,
                "templateId": d.templateId,
                "templateSettings": d.templateSettings or {},
                "sections": d.sections or [],
                "roleId": d.roleId,
                "branchOfVersionId": d.branchOfVersionId,
                "revision": d.revision,
                "createdAt": d.createdAt.isoformat() if d.createdAt else None,
                "updatedAt": d.updatedAt.isoformat() if d.updatedAt else None,
            }
            for d in drafts
        ]
    }

@router.post("/drafts", status_code=status.HTTP_201_CREATED)
@router.post("/v1/drafts", status_code=status.HTTP_201_CREATED)
def create_draft(
    payload: CreateDraftRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    draft = ResumeDraft(
        userId=user.id,
        title=payload.title or "Untitled Draft",
        templateId=payload.templateId or "classic",
        templateSettings=payload.templateSettings or {},
        sections=payload.sections or [],
        roleId=payload.roleId,
        branchOfVersionId=payload.branchOfVersionId,
    )
    db.add(draft)
    db.commit()
    db.refresh(draft)

    return {
        "status": "success",
        "data": {
            "id": draft.id,
            "title": draft.title,
            "templateId": draft.templateId,
            "templateSettings": draft.templateSettings or {},
            "sections": draft.sections or [],
            "roleId": draft.roleId,
            "branchOfVersionId": draft.branchOfVersionId,
            "revision": draft.revision,
            "createdAt": draft.createdAt.isoformat() if draft.createdAt else None,
            "updatedAt": draft.updatedAt.isoformat() if draft.updatedAt else None,
        }
    }

@router.get("/drafts/{draft_id}")
@router.get("/v1/drafts/{draft_id}")
def get_draft(
    draft_id: str,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    draft = db.query(ResumeDraft).filter(ResumeDraft.id == draft_id, ResumeDraft.userId == user.id).first()
    if not draft:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Draft not found.")

    return {
        "status": "success",
        "data": {
            "id": draft.id,
            "title": draft.title,
            "templateId": draft.templateId,
            "templateSettings": draft.templateSettings or {},
            "sections": draft.sections or [],
            "roleId": draft.roleId,
            "branchOfVersionId": draft.branchOfVersionId,
            "revision": draft.revision,
            "createdAt": draft.createdAt.isoformat() if draft.createdAt else None,
            "updatedAt": draft.updatedAt.isoformat() if draft.updatedAt else None,
        }
    }

@router.patch("/drafts/{draft_id}")
@router.patch("/v1/drafts/{draft_id}")
def update_draft(
    draft_id: str,
    payload: UpdateDraftRequest,
    user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    draft = db.query(ResumeDraft).filter(ResumeDraft.id == draft_id, ResumeDraft.userId == user.id).first()
    if not draft:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Draft not found.")

    if payload.title is not None:
        draft.title = payload.title
    if payload.templateId is not None:
        draft.templateId = payload.templateId
    if payload.templateSettings is not None:
        draft.templateSettings = payload.templateSettings
    if payload.sections is not None:
        draft.sections = payload.sections
    if payload.roleId is not None:
        draft.roleId = payload.roleId

    draft.revision += 1
    db.commit()
    db.refresh(draft)

    return {
        "status": "success",
        "data": {
            "id": draft.id,
            "title": draft.title,
            "templateId": draft.templateId,
            "templateSettings": draft.templateSettings or {},
            "sections": draft.sections or [],
            "roleId": draft.roleId,
            "revision": draft.revision,
            "updatedAt": draft.updatedAt.isoformat() if draft.updatedAt else None,
        }
    }
